Upgrade to support Xcode 14.0

mokagio opened the pull request:
Note: While I am listed as the PR author, @crazytonyli did all the heavy lifting to get the app building and up-to-date through all the betas! _—@mokagio_

---


### Description

This is going to be a long running branch to (try to) stay on top of the Xcode 14 betas PRs.

I expected many of you will want to checkout this branch locally to play with it. As such, I think we should keep it alive by _merging `trunk`_ into it. But, I think once we'll officially switch to Xcode 14 stable, it'll make sense to do a big rebase so the history stays tidy.

I called the branch `xcode-14`. This might give us the option to keep the beta branch stable (🙃 🙂) and work on a new beta support without breaking the "stable" branch. That is, we might have an `xcode-14-beta-2` branch in the future.

**Ideally, we should implement as many fixes for errors and warnings found here as possible on `trunk`.** See:

- #7057 

### Testing instructions

Download Xcode 14 beta 1 and run this branch 😅 .

If you do, you might notice a bunch of:

> Cannot access property 'cancellableSiteID' with a non-sendable type 'AnyCancellable?' from non-isolated deinit; this is an error in Swift 6

I haven't looked into it yet, other than finding [this thread about it in the Swift forums](https://forums.swift.org/t/cannot-access-property-with-a-non-sendable-type-from-non-isolated-deinit/57877).

### Screenshots

<table>
<tr>
	<td><img width="1579" alt="Screen Shot 2022-06-09 at 7 37 25 pm" src="https://user-images.githubusercontent.com/1218433/172820518-957247b8-a71e-46f4-a998-cf329a256edd.png"></td>
	<td><img width="561" alt="Screen Shot 2022-06-09 at 7 44 16 pm" src="https://user-images.githubusercontent.com/1218433/172820652-e0618679-4289-455a-84b3-ed1855c705cc.png"></td>
</table>

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

mokagio committed:
Upgrade WooFoundation to recommended Xcode 14.0 beta 1 settings

This also touched all the schemes to update the `LastUpgradeVersion` to
track they were checked in Xcode 14.0.

crazytonyli committed:
Merge branch 'trunk' into xcode-14

crazytonyli committed:
Fix a code signing issue on Xcode 14 beta 4

crazytonyli committed:
On iOS 16, use UIKit to hide or show navigation bar on the HubMenu

Fix https://github.com/woocommerce/woocommerce-ios/issues/7393

crazytonyli committed:
Assert the correct localized date string based on OS version

On iOS 16, `DateFormatter` returns a different localized string for template
format 'MMM d, h:mm a', compared to previous iOS versions.

Feeback to Apple: FB11335645.

crazytonyli committed:
Merge branch 'trunk' into xcode-14

crazytonyli committed:
Merge branch 'trunk' into xcode-14

crazytonyli committed:
Merge branch 'trunk' into xcode-14

crazytonyli committed:
Explicitly make amount text field lose keyboard focus

crazytonyli committed:
Merge branch 'trunk' into xcode-14

crazytonyli committed:
Use Xcode 14 on CI

crazytonyli committed:
Use iPhone 14 to run UI tests

crazytonyli committed:
Merge pull request #7371 from woocommerce/xcode-14-workarounds

Make the app compatible with Xcode 14

crazytonyli commented:
Hi @mokagio , do you think this PR is ready for review now?

mokagio commented:
> Hi @mokagio , do you think this PR is ready for review now?

Yes it is! Pushing the button and reaching out to the devs.

itsmeichigo reviewed (approved):
This works great for me both on the simulator and physical devices.
Thanks for the updates!

mokagio committed:
Upgrade project to Xcode 14.0 stable recommended settings

mokagio committed:
Upgrade CocoaPods to version 1.11.3

This seems to address the "Update project to recommended settings"
warning for the Pods project.

jaclync commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewController.swift:
```diff
@@ -31,6 +31,25 @@ final class HubMenuViewController: UIHostingController<HubMenu> {
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+
+        if #available(iOS 16.0, *) {
+            self.navigationController?.setNavigationBarHidden(false, animated: animated)
+        }
```
nit: does it make sense to reverse the order of the `super` call since this is a "disappear" event?

```suggestion       
        if #available(iOS 16.0, *) {
            self.navigationController?.setNavigationBarHidden(false, animated: animated)
        }
        
        super.viewWillDisappear(animated)
```

crazytonyli replied:
[The API doc](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621485-viewwilldisappear) says
> If you override this method, you must call super at some point in your implementation. 

I don't think the order of these two statements would make any difference, the pop animation should begin and end at the same time.

jaclync commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewController.swift:
```diff
@@ -31,6 +31,25 @@ final class HubMenuViewController: UIHostingController<HubMenu> {
+        // We want to hide navigation bar *only* on HubMenu screen. But on iOS 16, the `navigationBarHidden(true)`
+        // modifier on `HubMenu` view hides the navigation bar for the whole navigation stack.
+        // Here we manually hide or show navigation bar when entering or leaving the HubMenu screen.
+        if #available(iOS 16.0, *) {
```
do you still see the navigation bar issue https://github.com/woocommerce/woocommerce-ios/issues/7091 with the latest Xcode 14? somehow I don't see this issue anymore on `trunk` using Xcode 14 in an iOS 16.0 simulator. I don't have an iOS 16 device though, might be worth testing on one to confirm if this workaround is still needed.

ealeksandrov replied:
Just tried to comment out `viewWillAppear` and `viewWillDisappear` - reproduced the issue on the device:

Payments | Reviews
--|--
![IMG_2289](https://user-images.githubusercontent.com/3132438/190339977-86f58498-0437-4493-93e5-fcdc243b207b.PNG)|![IMG_2290](https://user-images.githubusercontent.com/3132438/190339994-10bb1349-f973-4e52-833b-49645306db80.PNG)

crazytonyli replied:
@jaclync Hmm, that's strange. I just run the trunk branch on an iPhone 14 simulator using Xcode 14, and verified this is still an issue.

jaclync replied:
Sounds like something odd on my end then, thanks for checking both!

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/ShippingLineDetails.swift:
```diff
@@ -52,6 +52,9 @@ struct ShippingLineDetails: View {
                                                  keyboardType: .default)
                                 .accessibilityIdentifier("add-shipping-name-field")
+                                .onTapGesture {
+                                    focusAmountInput = false
```
do you have the testing steps for this change?

crazytonyli replied:
This code was added so that the `test_creat_new_order` UI test can pass on CI. It's a strange issue though, since the test failure only occurs on CI, not on my mac.

BTW, the issue was mentioned in [my original PR](https://github.com/woocommerce/woocommerce-ios/pull/7371#issue-1320359748).

> Here is a very strange issue only happens in UI test on CI. I had to explicitly make [a text field](https://github.com/woocommerce/woocommerce-ios/pull/7371/files#diff-e1ee95623c8f8278d8814fed88be69d800c8ad94be3cbb963a1be585ad3fa979R35) loose keyboard focus by [changing the binding variable](https://github.com/woocommerce/woocommerce-ios/pull/7371/files#diff-e1ee95623c8f8278d8814fed88be69d800c8ad94be3cbb963a1be585ad3fa979R56), for the other "add-shipping-name-field" text field to gain keyboard focus.

crazytonyli replied:
BTW, you can find my many attempts of fixing this issue in [this buildkite branch builds](https://buildkite.com/automattic/woocommerce-ios/builds?branch=xcode-14-workarounds).

From the test result bundle (you can download one from the artifacts of [this step](https://buildkite.com/automattic/woocommerce-ios/builds/7715#01831096-c84d-40d0-b9ea-7d6307043512)), it appears even thought "add-shipping-name-field" is tapped, it's still "add-shipping-amount-field" that has the keyboard focus, which caused the test to fail.
<img width="545" alt="Screen Shot 2022-09-16 at 9 17 57 AM" src="https://user-images.githubusercontent.com/1101828/190510954-0139e842-ad06-4af9-94c9-b3e3c0557d54.png">

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Dashboard/Stats V4/StoreStatsPeriodViewModelTests.swift:
```diff
@@ -341,7 +341,11 @@ final class StoreStatsPeriodViewModelTests: XCTestCase {
 
         // Then
-        XCTAssertEqual(timeRangeBarViewModels.map { $0.timeRangeText }, ["Monday, Jan 3", "Monday, Jan 3, 1:00 AM"])
+        if #available(iOS 16.0, *) {
+            XCTAssertEqual(timeRangeBarViewModels.map { $0.timeRangeText }, ["Monday, Jan 3", "Monday, Jan 3 at 1:00 AM"])
+        } else {
+            XCTAssertEqual(timeRangeBarViewModels.map { $0.timeRangeText }, ["Monday, Jan 3", "Monday, Jan 3, 1:00 AM"])
+        }
```
nice catch!

crazytonyli committed:
Merge pull request #7725 from woocommerce/mokagio/update-project-settings

Upgrade project to Xcode 14.0 stable recommended settings

jaclync reviewed (approved):
Working as before on an iOS 15 device and iOS 16 simulator! Just some minor non-blocking questions

ealeksandrov reviewed (approved):
✅ Xcode 13.4.1 + iOS 14.5, 15.5 simulators
✅ Xcode 14 + iOS 16 simulator + iOS 16 device

There are couple of new warnings in Xcode 14:

```
warning build: Run script build phase 'SwiftLint' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase.


warning build: Run script build phase 'Check for nested frameworks' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase.
```

crazytonyli commented:
Hi @jaclync , I'll merge this PR for now. I'm happy to follow up if you have any further comments. Thanks! 😄
